### DISCUSSION TITLE
Another way of having a initial state?

### DISCUSSION THREAD

**Discussion by gmcamposano** [pmndrs/zustand#1622] [General]
Right now I have a local state that saves a total discounted value:
It searches the cart, which is coming from a zustand state, and sets it with reduce:
```
  useEffect(() => {
    if (!carrito.length) return
    setTotalDescontado(
      carrito.reduce(
        (acc, item) =>
          acc + item.precio * item.cantidad - +item.total.toFixed(2),
        0
      )
    )
  }, [carrito])
```
What I would like to do is store this

Instead of having this method, and having to call it every time using a useEffect, would it be possible to do it automatically without setting it? 
```
  setTotalDescontado: () =>
    set((state) => ({
      totalDescontado: state.carrito.reduce(
        (acc, item) =>
          acc + item.precio * item.cantidad - +item.total.toFixed(2),
        0
      )
    })),
```

Example: Something like this which does not work
```
const initial_state = {
  carrito: [],
  totalDescontado: carrito.reduce(
    (acc, item) =>
      acc + item.precio * item.cantidad - +item.total.toFixed(2),
    0
  )
}
```
This way the logic exists within the store and when I access the discounted value is already processed.

**Comment by dai-shi** [comment]
In general, it's a bad practice to use useEffect to derive state.

For deriving state in zustand, the common/typical way is to use selectors.
You could also compute derived state in actions, or reach out some middlewares to that helps the pattern. https://github.com/pmndrs/zustand/blob/main/docs/integrations/third-party-libraries.md
